IK-868: Documentation for gradients (#239)
imagekitio/docs-keep-a-changelog: features/image-transformations/image-enhancement-and-color-manipulation.md

{% endtab %}

{% endtabs %}

### Gradient - (e-gradient)

This is used to add a gradient overlay over an input image.

> The gradient formed is a *linear gradient* (containing two colors) and it can be customised with the following parameters:

Reviewer: ```suggestion
The gradient formed is a *linear gradient* containing two colors, and it can be customised with the following parameters:
```

> | Parameter       | Description       | Value Type        | Range      | Default Value     |

Reviewer: General point: In the `Default Value` column, can we use bulleted lists instead of just line breaks?

Reviewer 2: I had used bullets, but tbh line breaks look way better and cleaner in this particular case.

Reviewer: resolved

|-----------------|-------------------|-------------------|------------|-------------------|

> | Linear Direction (`ld`) | **Optional**.<br/><br/>This is used to specify the direction vector of the linear gradient.<br/><br/>For enhanced control over the direction, this parameter also accepts a integer `0-359`, which specifies the angle (in degrees) when measured over a standard x-y axes with pivot passing through the geometrical centre of the image. For example, `0` would mean `top`, `45` would mean `top_right`, `90` would mean `right` etc. | `Number` or `String` | **Number** - `0` to `359`<br/><br/>**String** - `top`, `top_right`, `right`, `bottom_right`, `bottom`, `bottom_left`, `left`, `topleft` | `180` (or `bottom`) |

Reviewer: Keeping in line with the descriptions for `e-shadow` parameters,

```suggestion
| Linear Direction (`ld`) | **Optional**.<br/><br/>Sets the direction vector of the linear gradient.<br/><br/>For enhanced control over the direction, this parameter also accepts an integer in the range `0-359`, which specifies the angle (in degrees) when measured over a standard x-y axes with pivot passing through the geometrical centre of the image. For example, `0` would mean `top`, `45` would mean `top_right`, `90` would mean `right` etc. | `Number` or `String` | **Number** - `0` to `359`<br/><br/>**String** - `top`, `top_right`, `right`, `bottom_right`, `bottom`, `bottom_left`, `left`, `topleft` | `180` (or `bottom`) |
```

> | From Color (`from`) | **Optional**.<br/><br/>This is used to specify a *from* color for the gradient.<br/><br/>RGBA Code (e.g. `FFAABB50`) is a color representation model where the first six characters must be the RGB Hex Code and the last two characters must be a number between `00` and `99`, which indicates the opacity level. For example, `00` represents an opacity level of `0.00`, `01` represents an opacity level of `0.01`, and so on. | `String` | An [SVG recognized color](https://www.w3.org/TR/SVG11/types.html#ColorKeywords) keyword name eg. `aqua`;<br/><br/>or a RGB Hex Code eg. `FF0000`;<br/><br/>or a RGBA Hex Code eg. `FF000040`; | `FFFFFF` |

Reviewer: `This is used to specify a *from* color for the gradient.`

This line isn't actually describing anything. Can it be rewritten to include more information about the parameter?

Reviewer 2: Any suggestions? I can't think of anything more that would be needed to expand this.

Reviewer: It can have the keywords `start` or `starting` maybe. Something like "the starting point of the gradient block will have this color in its purest form"

Reviewer 2: Done

Reviewer: resolved

> | To Color (`to`) | **Optional**.<br/><br/>This is used to specify a *to* color for the gradient.<br/><br/>RGBA Code (e.g. `FFAABB50`) is a color representation model where the first six characters must be the RGB Hex Code and the last two characters must be a number between `00` and `99`, which indicates the opacity level. For example, `00` represents an opacity level of `0.00`, `01` represents an opacity level of `0.01`, and so on. | `String` | An [SVG recognized color](https://www.w3.org/TR/SVG11/types.html#ColorKeywords) keyword name eg. `aqua`;<br/><br/>or a RGB Hex Code eg. `FF0000`;<br/><br/>or a RGBA Hex Code eg. `FF000040`; | `00000000` |

Reviewer: Same as https://github.com/imagekitio/docs-keep-a-changelog/pull/239/files#r1448978734

Reviewer 2: Done

Reviewer: resolved

> | Stop Point (`sp`) | **Optional**.<br/><br/>This is used to dictate where the break point should be made between the `from` color and the `to` color. In simpler words, it is the point (along the direction vector of the gradient) at which the `to` color achieves it’s pure form i.e. without any blends.<br/><br/>When the number is less than or equal to 1 i.e. a floating point, it would be treated as a percentage value. For example, `0.25` means 25%, `0.1` means 10% and so on.<br/><br/>When the number is greater than 1, it would be treated as an absolute value (in pixels). For example, `100` would mean 100px and so on. | `Number` or `String` | **Number** - A floating point number between `0.01` and `0.99` or an integer greater than or equal to `1`.<br/><br/>**String** - An arithmetic expression like `ih_div_2`. Read more about arithmetic expressions [here](../arithmetic-expressions-in-transformations.md). | `1` |

Reviewer: ```suggestion
| Stop Point (`sp`) | **Optional**.<br/><br/>This is used to dictate where the break point should be made between the `from` color and the `to` color. In simpler words, it is the point (along the direction vector of the gradient) at which the `to` color achieves it’s pure form i.e. no blending with the `from` color. <br/><br/>When the number is less than or equal to 1 i.e. a floating point, it would be treated as a percentage value. For example, `0.25` means 25%, `0.1` means 10% and so on.<br/><br/>When the number is greater than 1, it would be treated as an absolute value (in pixels). For example, `100` would mean 100px and so on. | `Number` or `String` | **Number** - A floating point number between `0.01` and `0.99` or an integer greater than or equal to `1`.<br/><br/>**String** - An arithmetic expression like `ih_div_2`. Read more about arithmetic expressions [here](../arithmetic-expressions-in-transformations.md). | `1` |
```

Multiple parameters can be configured by appending them with an underscore (`_`). They will be applied independent of the order in which they are specified.

Example usage - `e-gradient-<linear-direction>_<from-color>_<to-color>_<stop-point>`, which would look like, `e-gradient-ld-top_right_from-FF000040_to-00FF0040_sp-0.9`.

{% tabs %}

> {% tab title="Original Image" %}

Reviewer: `Original` is used at other places to mean the absolutely unprocessed original image. But we are resizing to 300 width here.

```suggestion
{% tab title="Base image" %}
```

Reviewer 2: I don't think that's true. I've seen multiple places to mean a "starting point" and not an "absolutely unprocessed" image. Ref: https://docs.imagekit.io/features/image-transformations/image-enhancement-and-color-manipulation

Reviewer: For `e-sharpen`, this title is used, which I like better.
`Original` is fine too though on second thought because quality and format optimizations always happen by default anyway. Both would be okay to go with.
 
<img width="817" alt="image" src="https://github.com/imagekitio/docs-keep-a-changelog/assets/19747413/b10cfbc5-ee95-4cf1-8bdb-ec6c06768263">

Reviewer 2: Cool, let's keep it as-is then.

Reviewer: resolved

{% endtab %}
{% endtabs %}

#### Gradient as a background

For input images containing a solid object over a transparent background, gradient can also be applied as an underlay.

##### Usage syntax

```markup
https://ik.imagekit.io/<imagekitId>/your-image.png?tr=e-gradient-from-red_to-white:l-image,i-your-image.png,t-false,l-end
```

##### Examples

{% tabs %}
{% tab title="Original Image" %}
URL - [https://ik.imagekit.io/demo/tr:w-500/lipstick_2M.png](https://ik.imagekit.io/demo/tr:w-500/lipstick_2M.png)

![](https://ik.imagekit.io/demo/tr:w-500/lipstick_2M.png)
{% endtab %}

{% tab title="Gradient as a background" %}
URL - [https://ik.imagekit.io/demo/tr:e-gradient-from-lightskyblue_to-mintcream:l-image,i-lipstick_2M.png,t-false,l-end:w-500/lipstick_2M.png](https://ik.imagekit.io/demo/tr:e-gradient-from-lightskyblue_to-mintcream:l-image,i-lipstick_2M.png,t-false,l-end:w-500/lipstick_2M.png)

![](https://ik.imagekit.io/demo/tr:e-gradient-from-lightskyblue_to-mintcream:l-image,i-lipstick_2M.png,t-false,l-end:w-500/lipstick_2M.png)
{% endtab %}
{% endtabs %}

{% hint style="info" %}

> If you experience a positional displacement between the original image and the overlay image, it would be worthwhile to use `t-false` to disable default trimming of the overlay layer like `l-image,i-your-image.png,t-false,l-end`.

Reviewer: ```suggestion
If you experience a positional displacement between the original image and the overlay image, use `t-false` to disable default trimming of the overlay layer like `l-image,i-your-image.png,t-false,l-end`.
```

{% endhint %}

> #### Adding gradient blocks over images

Reviewer: We should add a section with a heading in the https://docs.imagekit.io/features/image-transformations/overlay-using-layers page too. A short description along with a link to this section can be added there.

Reviewer 2: I confirmed this with @manu4543 if we should add a section or not - it was decided to add a sentence in the solid color section which directs the user to gradient section in image manipulation section.

Reviewer: resolved

> You can also add a gradient block over an input image.

Reviewer: ```suggestion
Like solid color blocks ${link_to_page}, you can also add a gradient block of fixed dimensions over an input image.
```

Reviewer 2: There's a separate sentence towards the end which mentions this.
Not needed here, also because a reader has no context to "adding solid color blocks" if he/she lands here directly. So starting the section with a non-contextual information doesn't make much sense.

Reviewer: resolved

##### Usage syntax

```markup
https://ik.imagekit.io/<imagekitId>/your-image.png?tr=l-image,i-ik_canvas,e-gradient,l-end
```

You can also control the position of the gradient color overlay using these [positional parameters](./overlay-using-layers.md#position-of-layer).

##### Transformation of gradient block overlay

Following transformation parameters are supported on the gradient block overlay inside a layer.

> | Parameter                                                                                | Description                        |

Reviewer: `lx`, `ly`, `lfo` can also be added to this table

Reviewer 2: The pattern we're following has a separate descriptive sentence which covers positional parameters and has a link to the page. Adding it here is simply redundant.

Reviewer: resolved

|------------------------------------------------------------------------------------------|------------------------------------|

> | w                                                                                        | Width of solid color block. It can also accept arithmetic expressions such as `bw_div_2`, or `bw_mul_0.8`. Learn more about arithmetic expressions [here](../arithmetic-expressions-in-transformations.md). |

Reviewer: ```suggestion
| w                                                                                        | Width of the gradient block. It can also accept arithmetic expressions such as `bw_div_2`, or `bw_mul_0.8`. Learn more about arithmetic expressions [here](../arithmetic-expressions-in-transformations.md). |
```

> | h                                                                                        | Height of solid color block. It can also accept arithmetic expressions such as `bh_div_2`, or `bh_mul_0.8`. Learn more about arithmetic expressions [here](../arithmetic-expressions-in-transformations.md).  |

Reviewer: ```suggestion
| h                                                                                        | Height of the gradient block. It can also accept arithmetic expressions such as `bh_div_2`, or `bh_mul_0.8`. Learn more about arithmetic expressions [here](../arithmetic-expressions-in-transformations.md).  |
```
